generators/CuraPackage.py

```diff
@@ -0,0 +1,399 @@
+
+    :param directory_path: The path to a directory to get the hash of.
+    :return: A cryptographic hash of the specified directory.
+    """
+    hash_list: List[Tuple[str, str]] = []  # Contains a list of (relative_file_path, file_hash) tuples
+    for root, _, filenames in os.walk(directory_path):
+        if root in exclude_paths:
+            continue
+        for filename in filenames:
+            rel_dir_path = os.path.relpath(root, directory_path)
+            rel_path = os.path.join(rel_dir_path, filename)
+            if rel_dir_path in exclude_paths or rel_path in exclude_paths:
+                continue
+            abs_path = os.path.join(root, filename)
+            hash_list.append((rel_path, _hashFile(abs_path, verbose)))
+
+    # We need to be sure that the list is sorted by the relative_file_path to account for cases where the files are read
+    # in different orders.
+    ordered_list = sorted(hash_list, key=lambda x: x[0])
+
+    hasher = hashlib.sha256()
+    for i in ordered_list:
+        hasher.update("".join(i).encode('utf-8'))
+    return hasher.hexdigest()
+
+
+class CuraPackage(Generator):
+    _content_types_xml = r"""<?xml version="1.0" encoding="UTF-8"?>
+<Types xmlns="http://schemas.openxmlformats.org/package/2006/content-types">
+  <Default Extension="rels" ContentType="application/vnd.openxmlformats-package.relationships+xml" />
+  <Default Extension="xml.fdm_material" ContentType="application/x-ultimaker-material-profile" />
+  <Default Extension="xml.fdm_material.sig" ContentType="application/x-ultimaker-material-sig" />
+  <Default Extension="inst.cfg" ContentType="application/x-ultimaker-quality-profile" />
+  <Default Extension="def.json" ContentType="application/x-ultimaker-machine-definition" />
+  <Default Extension="json" ContentType="text/json" />
+</Types>
+"""
+
+    _dot_rels = r"""<?xml version="1.0" encoding="UTF-8"?>
+<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
+  <Relationship Target="/package.json" Type="http://schemas.ultimaker.org/package/2018/relationships/opc_metadata" Id="rel0" />
+</Relationships>"""
+
+    _package_json_rels = Template(r"""<?xml version="1.0" encoding="UTF-8"?>
+<Relationships xmlns="http://schemas.openxmlformats.org/package/2006/relationships">
+  <Relationship Target="/files/{{ package_type }}s" Type="{{ package_type }}" Id="rel0" />
+</Relationships>""")
+
+    _deps_init_py = Template(r"""def initialize_paths():
+    from UM.Logger import Logger
+    Logger.info("Initializing the correct paths for {{ package_id }}")
+    
+    from pathlib import Path
+    from platform import system
+    from sys import path
+    import sysconfig
+    
+    file_manifest = None
+    {% if central_storage %}# Read the Central Storage manifest
+    from UM.CentralFileStorage import CentralFileStorage
+    import json
+    central_storage_file = Path(__file__).parent.parent.joinpath("central_storage.json")
```

Comment:
Would be good to document such stuff, e.g. with an example of the directory layout.

Reply:
Yes it would be, but that's a bit outside of the scope of this ticket, I think.

I'll make a technical ticket on the backlog together with the 'replace standard-project cmake file with conan-generated one'.